### ISSUE TITLE
Force password change

### ISSUE THREAD

**Issue by jeremy-walker** [ltsp/ltsp#148]
Let's say the administrator creates a new user account(s) with a temporary password on the LTSP server, and then provides the credentials to the user. How can we now force the user to change the temporary password when the user logs in to the LTSP client? 

I tried expiring the password of the new user, but that results in a login failed on the client. LTSP clients are running Ubuntu Mate 20.04.

**Comment by alkisg** [comment]
Authentication in Linux is handled by PAM. PAM allows e.g. the display manager to do a "conversation" with the backend, so that messages like "your password has expired, please enter a new one" are possible.

Currently, LTSP is using the pam_exec module to hook into PAM without using compiled .c code. This allows us to get the username and password and test it via SSH against the LTSP server, but AFAIK pam_exec doesn't allow the script (=LTSP) to do conversation with the display manager. This means that with the current approach, SSH conversation cannot be supported.

So the proper solution would be for someone to develop and maintain a PAM module, written in .c, for SSH-based authentication. I have no plans to do that, unless a lot of funding is found, so I'll mark this issue as "won't fix".

People that need this functionality may use LDAP instead, which ships with a proper .c module and can do conversation.

**Comment by jeremy-walker** [comment]
Hi @alkisg, thanks for the clarification. 

Just to confirm, this means that users cannot change their passwords (using LTSP clients) at all, correct? In other words, in the default (non LDAP) LTSP setup, the only 2 ways a user's password can be changed is:

1. admin changes it for the user
2. user logs in to the server via ssh or some other method, and changes the password on the server

I am trying to figure out if there is a workable alternative to installing /setting up LDAP. If there isn't one, I will look into setting up LDAP since our policies require users to periodically change their passwords.

**Comment by alkisg** [comment]
The users were never able to change their passwords locally with LTSP.

With the old LTSP, people were running something like `ltsp-remoteapps gnome-about-me`.
With the new LTSP, ltsp-remoteapps do not exist, but it's easy to write a script that runs `ssh -X server gnome-about-me`, while automatically configuring and using passwordless SSH (with keys).

**Comment by jeremy-walker** [comment]
Thanks. Since my server is running on a hardware different from my clients, I'm running it headless, so no `mate-about-me` available on the server. In any case, ldap is useful to us in many other ways. I will start looking into setting up openldap to work with LTSP.

**Comment by alkisg** [comment]
You can have xorg installed on a headless server; `ssh -X` will work fine even if it itself doesn't run xorg.

You can also write a tiny zenity or python/gtk script that gets the new and old passwords, and screen-scrapes an `ssh server passwd` command.

But sure, LDAP can do a lot more as it comes with a real PAM module.

**Comment by alkisg** [comment]
Some notes on how non-LDAP tools behave regarding expired passwords.

To expire a password, the sysadmin can run: `passwd --expire john`
At the next `login` (or `su - john`), the user will be prompted to change it:

```
$ login john
Password: 
You are required to change your password immediately (administrator enforced)
Changing password for john.
Current password: ********
New password: ********
Retype new password: ********
Welcome...
```

But this doesn’t work properly with SSH: it allows you to login, then prompts for a password change, then forcefully logs you out:

```
$ ssh john@localhost
john@localhost's password: ********
You are required to change your password immediately (administrator enforced)
Welcome...
Changing password for john.
Current password: ********
New password: ********
Retype new password: ********
passwd: updated successfully
Connection to localhost closed.
```

It’s worse with the SSH SFTP subsystem that LTSP uses, and with SSHFS, as they don’t even allow the user to login, with no messages:

```
$ sshfs john@localhost: john
john@localhost's password: 
remote host has disconnected
```

To sum up, SSH doesn't handle password expiration and renewal properly, so they can't be implemented in pamltsp as it relies on SSH.

People that need them should use LDAP instead, see #338.
In that case, pam_sss.so will hopefully handle password renewal properly. Then pamltsp will be called, and it will use the updated password to connect to the SSHFS server with no issues.
